cmd/gateway/s3x/gateway-s3x.go

```diff
@@ -95,26 +120,75 @@ func init() {
 
 func temxGatewayMain(ctx *cli.Context) {
 	minio.StartGateway(ctx, &TEMX{
-		HTTPAddr: ctx.String("info.http.endpoint"),
-		GRPCAddr: ctx.String("info.grpc.endpoint"),
-		DSPath:   ctx.String("ds.path"),
-		XAddr:    ctx.String("temporalx.endpoint"),
-		Insecure: ctx.Bool("temporalx.insecure"),
+		HTTPAddr:  ctx.String("info.http.endpoint"),
+		GRPCAddr:  ctx.String("info.grpc.endpoint"),
+		DSPath:    ctx.String("ds.path"),
+		DSType:    DSType(ctx.String("ds.type")),
+		CrdtTopic: ctx.String("ds.topic"),
+		XAddr:     ctx.String("temporalx.endpoint"),
+		Insecure:  ctx.Bool("temporalx.insecure"),
 	})
 }
 
-// newLedgerStore returns an instance of ledgerStore that uses badgerv2
-func (g *TEMX) newLedgerStore(dsPath string, dag pb.NodeAPIClient) (*ledgerStore, error) {
+// newLedgerStore returns an instance of ledgerStore
+func (g *TEMX) newLedgerStore(ctx context.Context, dag pb.NodeAPIClient, pub pb.PubSubAPIClient) (*ledgerStore, error) {
+	switch g.DSType {
+	case DSTypeBadger:
+		return g.newBadgerLedgerStore(dag)
+	case DSTypeCrdt:
+		return g.newCrdtLedgerStore(ctx, dag, pub)
+	}
+	return nil, fmt.Errorf(`data store type "%v" not supported`, g.DSType)
+}
+
+// newBadgerLedgerStore returns an instance of ledgerStore that uses badgerv2
+func (g *TEMX) newBadgerLedgerStore(dag pb.NodeAPIClient) (*ledgerStore, error) {
 	opts := badger.DefaultOptions
-	ds, err := badger.NewDatastore(dsPath, &opts)
+	ds, err := badger.NewDatastore(g.DSPath, &opts)
 	if err != nil {
 		return nil, err
 	}
 	return newLedgerStore(ds, dag)
 }
 
+// newCrdtLedgerStore returns an instance of ledgerStore that uses crdt and backed by badgerv2
+func (g *TEMX) newCrdtLedgerStore(ctx context.Context, dag pb.NodeAPIClient, pub pb.PubSubAPIClient) (*ledgerStore, error) {
+	store, err := badger.NewDatastore(g.DSPath, &badger.DefaultOptions)
+	if err != nil {
+		return nil, err
+	}
+	//from the doc: The broadcaster can be shut down by canceling the given context. This must be done before Closing the crdt.Datastore, otherwise things may hang.
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		if cancel != nil {
+			cancel()
+		}
+	}()
```

Comment:
i think this might be whats causing the error??? anytime you create a crdt store here, you will be cancelling this context?

Reply:
That cancellation is from timeout (see `300.05s`)

That cancel is only called if there is an early error exist.

Reply:
ah okay i see it my mistake